### PR TITLE
Fix #4886 look for a tarball first fallback to individual files

### LINKED ISSUE #4886: SRSClient download bug for 2020 dates
### Description

Try to download some SRS files from the year 2020 which appears to work but no files are downloaded. Looking at the source for the data the 2020 tarball has not been create yet which is the root cause. Proposal from @Cadair was to try tarball first and fallback to daily files. Should try to ensure if there is a failure in the download process we don't return 👻  files 

### Expected vs Actual behaviour
The SRS files to be downloaded or an error raised. Currently it fails silently and even gives back a list of file paths that don't exist.

### Steps to Reproduce

```python
from sunpy.net import Fido, attrs as a
results = Fido.search(a.Time('2020-01-01', '2020-01-20'), a.Instrument.srs_table)
dl = Fido.fetch(results)
# none of the returned  paths exist
```

### System Details

- SunPy Version: 2.0.5
- Astropy Version:
- Python Version:
- OS information:

### PR DESCRIPTION
**Pull request by samaloney**
### Description

Fixes #4886 by checking for a yearly tarball first if not found, try individual SRS files.

### CODE REVIEW

File: sunpy/net/dataretriever/sources/tests/test_noaa.py [hunk 1]
```diff
@@ -175,6 +176,13 @@ def test_srs_tar_unpack_midyear():
     assert res.data[-1].endswith("20110608SRS.txt")
 
 
+@pytest.mark.remote_data
+def test_srs_missing_tarball():
```
Comment (nabobalis) [inline sunpy/net/dataretriever/sources/tests/test_noaa.py:180]: What is this testing?

Comment (samaloney) [reply sunpy/net/dataretriever/sources/tests/test_noaa.py:180]: Yea I'm not sure the best approach for this test - as of now there is no tarball for 2020 so this test used to fail but not passed. But at some stage I assume noaa will create the 2020 tarball then this test is kinda obsolete.

File: sunpy/util/scraper.py [hunk 2]
```diff
@@ -358,17 +359,17 @@ def _smallerPattern(self, directoryPattern):
...
-                return TimeDelta(1*u.minute)
+                return relativedelta(minutes=1)
             elif any(hour in directoryPattern for hour in ["%H", "%I"]):
-                return TimeDelta(1*u.hour)
+                return relativedelta(hours=1)
             elif any(day in directoryPattern for day in ["%d", "%j"]):
-                return TimeDelta(1*u.day)
+                return relativedelta(days=1)
             elif any(month in directoryPattern for month in ["%b", "%B", "%m"]):
-                return TimeDelta(31*u.day)
```
Comment (samaloney) [inline sunpy/util/scraper.py:369]: I really don't understand how this used to work as not all months have 31 days?

File: sunpy/net/dataretriever/sources/noaa.py [hunk 3]
```diff
@@ -139,35 +143,70 @@ class SRSClient(GenericClient):
     2 Results from the SRSClient:
          Start Time           End Time      Instrument Physobs Source Provider
     ------------------- ------------------- ---------- ------- ------ --------
-    2016-01-01 00:00:00 2016-12-31 23:59:59       SOON     SRS   SWPC     NOAA
```
Comment (samaloney) [inline sunpy/net/dataretriever/sources/noaa.py:142]: Yea this docstring test was just wrong

File: sunpy/util/scraper.py (hunk 2 above)
Comment (Cadair) [reply sunpy/util/scraper.py:369]: yeah...

File: sunpy/net/dataretriever/sources/tests/test_noaa.py (hunk 1 above)
Comment (Cadair) [reply sunpy/net/dataretriever/sources/tests/test_noaa.py:181]: You could make this so it's the current year - 1? Then at least if this code breaks it will fail in Jan 2022 :rofl: 

I think the only sensible way to do this test is mock the responses from the server?

File: changelog/4904.bugfix.rst [hunk 4]
```diff
@@ -0,0 +1,3 @@
+Fix `SRSClient` which silenlty failed to download the SRS files when the tarball for the previous
```
Comment (Cadair) [inline changelog/4904.bugfix.rst:1]:
```suggestion
Fix `~.SRSClient` which silently failed to download the SRS files when the tarball for the previous
```

File: sunpy/net/dataretriever/sources/noaa.py [hunk 5]
```diff
@@ -148,37 +151,79 @@ class SRSClient(GenericClient):
     Results from 1 Provider:
     <BLANKLINE>
     2 Results from the SRSClient:
+<<<<<<< HEAD
```
Comment (Cadair) [inline sunpy/net/dataretriever/sources/noaa.py:154]: whoops

File: sunpy/net/dataretriever/sources/noaa.py [hunk 6]
```diff
@@ -148,37 +151,79 @@ class SRSClient(GenericClient):
...
+        tarball_urls = dict()
+        tarball_scraper = Scraper(self.BASE_URL + '%Y/%Y_SRS.tar.gz')
+        tarballs = tarball_scraper.filelist(tarball_timerange)
+        max_tarball_year = None
+        for tb_url in tarballs:
+            date = tarball_scraper._extractDateURL(tb_url)
+            year = date.to_datetime().year
+            max_tarball_year = year
+            tarball_urls[year] = tb_url
+            log.debug('Tarball found for year %d', year)
```
Comment (Cadair) [inline sunpy/net/dataretriever/sources/noaa.py:200]:
```suggestion
            log.debug('SRS tarball found for year %d', year)
```

File: sunpy/net/dataretriever/sources/noaa.py [hunk 7]
```diff
@@ -148,37 +151,79 @@ class SRSClient(GenericClient):
...
+
+        # Create a new time range for the times not covered by tarballs, have to assume tarballs
+        # cover a year, and look for individual srs file for this time range.
+        srs_urls = dict()
+        min_file_year = max_tarball_year if max_tarball_year else start_year
+        min_file_date = datetime(min_file_year+1, 1, 1, 0, 0, 0)
```
Comment (Cadair) [inline sunpy/net/dataretriever/sources/noaa.py:206]:
```suggestion
        min_file_date = datetime(min_file_year + 1, 1, 1, 0, 0, 0)
```

File: sunpy/net/dataretriever/sources/noaa.py [hunk 8]
```diff
@@ -148,37 +151,79 @@ class SRSClient(GenericClient):
...
+                date = srsfile_scraper._extractDateURL(srs_url)
+                srs_urls[(date.datetime.year, date.datetime.month, date.datetime.day)] = srs_url
+                log.debug('SRS file found for year %d', date)
+
+        # Now iterate over all days and if the day is in a year we have a tarball for or a day there
+        # is a individual srs file add to the result with corresponding extdict
+        for day in timerange.get_dates():
+            day_ymd = (day.datetime.year, day.datetime.month, day.datetime.day)
+            extdict = {name: getattr(day.datetime, name) for name in ['year', 'month', 'day']}
+            if 1996 < day_ymd[0] <= cur_year:
```
Comment (Cadair) [inline sunpy/net/dataretriever/sources/noaa.py:221]: What's special about 1996 here again?

File: sunpy/net/dataretriever/sources/noaa.py [hunk 9]
```diff
@@ -148,37 +151,79 @@ class SRSClient(GenericClient):
...
+            for srs_url in srsfiles:
+                date = srsfile_scraper._extractDateURL(srs_url)
+                srs_urls[(date.datetime.year, date.datetime.month, date.datetime.day)] = srs_url
+                log.debug('SRS file found for year %d', date)
+
+        # Now iterate over all days and if the day is in a year we have a tarball for or a day there
+        # is a individual srs file add to the result with corresponding extdict
+        for day in timerange.get_dates():
+            day_ymd = (day.datetime.year, day.datetime.month, day.datetime.day)
+            extdict = {name: getattr(day.datetime, name) for name in ['year', 'month', 'day']}
```
Comment (Cadair) [inline sunpy/net/dataretriever/sources/noaa.py:220]: Would this work?
```suggestion
            extdict = {name: segment for segment in day_ymd}
```

**Review by Cadair — CHANGES_REQUESTED** [review]

File: sunpy/net/dataretriever/sources/noaa.py (hunk 8 above)
Comment (samaloney) [reply sunpy/net/dataretriever/sources/noaa.py:221]: before 1996 they report are hand written so can't be parsed I should have replaced this with `self.MIN_YEAR` anyway

File: sunpy/net/dataretriever/sources/noaa.py (hunk 9 above)
Comment (samaloney) [reply sunpy/net/dataretriever/sources/noaa.py:213]: No as name isn't defined then, I could make `day_ymd` a dictionary to start but the lookups in the loop get messy

File: sunpy/net/dataretriever/sources/tests/test_noaa.py (hunk 1 above)
Comment (samaloney) [reply sunpy/net/dataretriever/sources/tests/test_noaa.py:181]: I'll see how much trouble mocking the responses will be

Comment (samaloney) [reply sunpy/net/dataretriever/sources/tests/test_noaa.py]: Done

**Comment by samaloney** [comment]
I expect  the `test_srs_current_year` to fail as the SRS files for the first 5 day of 2021 are malformed
```
ftp> ls
229 Entering Extended Passive Mode (|||49316|)
150 Opening ASCII mode data connection for file list
-rw-r--r--   1 ftp      ftp           643 Jan  4 17:37 0101SRS.txt
-rw-r--r--   1 ftp      ftp           643 Jan  4 17:37 0102SRS.txt
-rw-r--r--   1 ftp      ftp           644 Jan  4 17:37 0103SRS.txt
-rw-r--r--   1 ftp      ftp           577 Jan  4 17:37 0104SRS.txt
-rw-r--r--   1 ftp      ftp           577 Jan  5 00:30 20210105SRS.txt
-rw-r--r--   1 ftp      ftp           563 Jan  6 00:30 20210106SRS.txt
```

**Comment by nabobalis** [comment]
> I expect the `test_srs_current_year` to fail as the SRS files for the first 5 day of 2021 are malformed
> 
> ```
> ftp> ls
> 229 Entering Extended Passive Mode (|||49316|)
> 150 Opening ASCII mode data connection for file list
> -rw-r--r--   1 ftp      ftp           643 Jan  4 17:37 0101SRS.txt
> -rw-r--r--   1 ftp      ftp           643 Jan  4 17:37 0102SRS.txt
> -rw-r--r--   1 ftp      ftp           644 Jan  4 17:37 0103SRS.txt
> -rw-r--r--   1 ftp      ftp           577 Jan  4 17:37 0104SRS.txt
> -rw-r--r--   1 ftp      ftp           577 Jan  5 00:30 20210105SRS.txt
> -rw-r--r--   1 ftp      ftp           563 Jan  6 00:30 20210106SRS.txt
> ```

Can we get this fixed?

File: sunpy/net/dataretriever/sources/noaa.py [hunk 10]
```diff
@@ -148,61 +149,80 @@ class SRSClient(GenericClient):
...
-            url = base_url + suffix
-            result.append(url)
-        return result
+        # Validate time range and return early if out side min/max
+        cur_year = Time.now().datetime.year
+        req_start_year = timerange.start.datetime.year
+        req_end_year = timerange.end.datetime.year
+        start_year = req_start_year if self.MIN_YEAR < req_start_year <= cur_year else None
+        end_year = req_end_year if self.MIN_YEAR < req_end_year <= cur_year else None
+        if start_year is None and end_year is None:
```
Comment (Cadair) [inline sunpy/net/dataretriever/sources/noaa.py:174]: If the start time or end time is out of range, but not both, then the start_year or end_year variables would be `None` and would presumably fail later in this method? Do we need to clip them to the MIN_YEAR / MAX_YEAR?

**Review by Cadair — CHANGES_REQUESTED** [review]
Thanks a lot for this, it looks good. Does this fix the representation of the results in the table, or does it not have any affect on that?

I think there is one issue with your logic, but otherwise looks good.

File: sunpy/net/dataretriever/sources/noaa.py (hunk 10 above)
Comment (samaloney) [reply sunpy/net/dataretriever/sources/noaa.py]: Yea that didn't catch that kind of edge case reworked and added new test for these cases.

File: setup.cfg [hunk 11]
```diff
@@ -171,6 +171,7 @@ filterwarnings =
     # an astropy table inside a test, which is generally only done when
     # debugging.
     ignore:the imp module is deprecated in favour of importlib:DeprecationWarning:ipykernel.iostream
+    ignore:The `use_readline` parameter:DeprecationWarning:
```
Comment (nabobalis) [inline setup.cfg:174]: Where does this come from?

File: sunpy/net/dataretriever/sources/noaa.py [hunk 12]
```diff
@@ -3,19 +3,22 @@
...
+from sunpy import log
 from sunpy.net import attrs as a
 from sunpy.net.dataretriever import GenericClient, QueryResponse
 from sunpy.time import TimeRange
 from sunpy.util.parfive_helpers import Downloader
+from sunpy.util.scraper import Scraper
 
 __all__ = ['NOAAIndicesClient', 'NOAAPredictClient', 'SRSClient']
 
+from sunpy.util.scraper import Scraper
```
Comment (nabobalis) [inline sunpy/net/dataretriever/sources/noaa.py:20]:
```suggestion
```

File: setup.cfg (hunk 11 above)
Comment (samaloney) [reply setup.cfg]: Accidentally committed change I need for PyCharm and pytest and ipthon shell to work. I've removed it

File: sunpy/util/scraper.py [hunk 13]
```diff
@@ -117,18 +118,59 @@ def range(self, timerange):
...
+                            orig_time_tup.tm_hour, 0, 0)
+        elif timestep == relativedelta(days=1):
+            new_time_tup = (orig_time_tup.tm_year, orig_time_tup.tm_mon, orig_time_tup.tm_mday,
+                            0, 0, 0)
+        elif timestep == relativedelta(months=1):
+            new_time_tup = (orig_time_tup.tm_year, orig_time_tup.tm_mon, 1, 0, 0, 0)
+        elif timestep == relativedelta(years=1):
+            new_time_tup = (orig_time_tup.tm_year, 1, 1, 0, 0, 0)
+
+        return datetime(*new_time_tup)
```
Comment (Cadair) [inline sunpy/util/scraper.py:172]: Reduce some duplication?
```suggestion
        time_tup = [orig_time_tup.tm_year, orig_time_tup.tm_mon, orig_time_tup.tm_mday,
                    orig_time_tup.tm_hour, orig_time_tup.tm_min, orig_time_tup.tm_sec]
        if timestep == relativedelta(minutes=1):
            time_tup[-1] = 0
        elif timestep == relativedelta(hours=1):
            time_tup[-2:] = [0, 0]
        elif timestep == relativedelta(days=1):
            time_tup[-3:] = [0, 0, 0]
        elif timestep == relativedelta(months=1):
            time_tup[-4:] = [1, 0, 0, 0]
        elif timestep == relativedelta(years=1):
            time_tup[-5:] = [1, 1, 0, 0, 0]

        return datetime(*time_tup)
```

File: sunpy/util/scraper.py [hunk 14]
```diff
@@ -117,18 +118,59 @@ def range(self, timerange):
...
+        timestep : `dateutil.relativedelta.relativedelta`
+            The smallest time step to floor
+        Returns
+        -------
+        `datetime.datetime`
+            The time floored at the given time step
+
+        """
+        date_parts = [int(p) for p in date.strftime('%Y,%m,%d,%H,%M,%S').split(',')]
+        date_parts[-1] = date_parts[-1] % 60
```
Comment (Cadair) [inline sunpy/util/scraper.py:152]: So what's the damage with regards to leap seconds here. I assume it doesn't really matter, but what if someone has a URL with a 61st second in a minute in it?

File: sunpy/net/dataretriever/sources/noaa.py [hunk 15]
```diff
@@ -150,59 +151,83 @@ class SRSClient(GenericClient):
...
+            max_tarball_year = year
+            tarball_urls[year] = tb_url
+            log.debug('SRS tarball found for year %d', year)
+
+        # Create a new time range for the times not covered by tarballs, have to assume tarballs
+        # cover a year, and look for individual srs file for this time range.
+        srs_urls = dict()
+        min_file_year = max_tarball_year if max_tarball_year else start_year
+        min_file_date = datetime(max_tarball_year, 12, 31, 23, 59, 59) if max_tarball_year else \
+            datetime(start_year, 1, 1, 0, 0, 0)
```
Comment (Cadair) [inline sunpy/net/dataretriever/sources/noaa.py:200]: Line continuation characters are evil ;)
```suggestion
        min_file_date = (datetime(max_tarball_year, 12, 31, 23, 59, 59) if max_tarball_year else
                         datetime(start_year, 1, 1, 0, 0, 0))
```

**Review by Cadair — COMMENTED** [review]
Wow this really spiralled didn't it.

I think it looks good, just a couple of small style comments. Has this grown to the point where it's no longer a simple bug fix and is really a behaviour change? (i.e. should this still get a 2.0 backport?)

File: sunpy/util/scraper.py [hunk 16]
```diff
@@ -152,24 +152,20 @@ def _date_floor(date, timestep):
         date_parts[-1] = date_parts[-1] % 60
         date = datetime(*date_parts)
         orig_time_tup = date.timetuple()
-        if timestep == relativedelta(seconds=1):
```
Comment (Cadair) [inline sunpy/util/scraper.py]: note that I deleted this branch, it's now the default. Do you think this is an issue?

Comment (samaloney) [reply sunpy/util/scraper.py]: Yea that will still work

File: sunpy/util/scraper.py (hunk 14 above)
Comment (nabobalis) [reply sunpy/util/scraper.py:152]: I assume it doesn't work?

**Review by nabobalis — APPROVED** [review]

File: sunpy/util/scraper.py (hunk 14 above)
Comment (samaloney) [reply sunpy/util/scraper.py:152]: Yea if you are actually looking for a file with a 60th second you won't find it even if it was there,  I think that's a pretty extreme edge. I don't even think it would have worked before either.

**Review by Cadair — APPROVED** [review]
